storage/pc-ata/ata.c

```diff
@@ -0,0 +1,533 @@
+
+	/* Send the command */
+	ata_writereg(base, REG_CMD, cmd, 1);
+
+	/* Do the transfer */
+	switch (cmd) {
+	case CMD_READ_PIO:
+	case CMD_READ_PIO_EXT:
+		ret = ata_pio(dev, nsectors, buff, DIR_READ);
+		break;
+
+	case CMD_WRITE_PIO:
+	case CMD_WRITE_PIO_EXT:
+		if ((ret = ata_pio(dev, nsectors, buff, DIR_WRITE)) < 0)
+			break;
+
+		/* Flush the hardware cache */
+		cmd = (dev->mode == AMODE_LBA48) ? CMD_CACHE_FLUSH_EXT : CMD_CACHE_FLUSH;
+		ata_writereg(base, REG_CMD, cmd, 1);
+		break;
+	}
+
+	return ret;
+}
+
+
+ssize_t ata_read(ata_dev_t *dev, offs_t offset, char *buff, size_t len)
+{
+	uint8_t cmd;
+	ssize_t ret;
+
+	switch (dev->mode) {
+	case AMODE_CHS:
+	case AMODE_LBA28:
+		cmd = CMD_READ_PIO;
+		break;
+
+	case AMODE_LBA48:
+		cmd = CMD_READ_PIO_EXT;
+		break;
+
+	default:
+		return -1;
+	}
+
+	mutexLock(dev->bus->lock);
+	ret = ata_access(dev, (uint64_t)(offset / dev->secsize), (uint16_t)(len / dev->secsize), cmd, (uint8_t *)buff);
+	mutexUnlock(dev->bus->lock);
+
+	return ret;
+}
+
+
+ssize_t ata_write(ata_dev_t *dev, offs_t offset, const char *buff, size_t len)
+{
+	uint8_t cmd;
+	ssize_t ret;
+
+	switch (dev->mode) {
+	case AMODE_CHS:
+	case AMODE_LBA28:
+		cmd = CMD_WRITE_PIO;
+		break;
+
+	case AMODE_LBA48:
+		cmd = CMD_WRITE_PIO_EXT;
+		break;
+
+	default:
+		return -1;
+	}
+
+	mutexLock(dev->bus->lock);
+	ret = ata_access(dev, (uint64_t)(offset / dev->secsize), (uint16_t)(len / dev->secsize), cmd, (uint8_t *)buff);
+	mutexUnlock(dev->bus->lock);
+
+	return ret;
+}
+
+
+static int ata_initdev(ata_bus_t *bus, ata_dev_t *dev, uint8_t devnum)
+{
+	void *base = bus->base;
+	uint8_t hcyl, lcyl, info[512];
+	uint16_t data;
+	int err, i;
+
+	dev->bus = bus;
+
+	/* Select the device */
+	ata_select(dev, devnum, 0, 0, -1);
+
+	/* Floating bus check */
+	if (ata_readreg(base, REG_STATUS, 1) == 0xFF)
+		return -ENXIO;
+
+	/* Send identify command */
+	ata_writereg(base, REG_CMD, CMD_IDENTIFY, 1);
+
+	/* Status = 0 => device doesn't exist */
+	if (!ata_readreg(base, REG_STATUS, 1))
+		return -ENXIO;
+
+	lcyl = (uint8_t)(ata_readreg(base, REG_LCYLINDER, 1));
+	hcyl = (uint8_t)(ata_readreg(base, REG_HCYLINDER, 1));
+
+	/* Detect device type */
+	if (!lcyl && !hcyl) {
+		dev->type = ATA;
+	}
+	else if (lcyl == 0x3C && hcyl == 0xC3) {
+		dev->type = SATA;
+	}
+	else if (lcyl == 0x14 && hcyl == 0xEB) {
+		dev->type = ATAPI;
+		/* Send identify packet command */
+		ata_writereg(base, REG_CMD, CMD_IDENTIFY_PACKET, 1);
+	}
+	else if (lcyl == 0x69 && hcyl == 0x96) {
+		dev->type = SATAPI;
+		/* Send identify packet command */
+		ata_writereg(base, REG_CMD, CMD_IDENTIFY_PACKET, 1);
+	}
+	else {
+		/* Unknown device */
+		return -ENXIO;
+	}
+
+	/* Wait until BSY clears and DRQ sets */
+	if ((err = ata_wait(bus, STATUS_BSY, STATUS_DRQ)))
+		return err;
+
+	/* Read device info */
+	for (i = 0; i < 256; i++) {
+		data = (uint16_t)ata_readreg(base, REG_DATA, 2);
+		((uint16_t *)info)[i] = ((data & 0xFF) << 8) | ((data >> 8) & 0xFF);

```

Comment:
You can cause alignment issue by casting char * to bigger sizeof type. You need to changed endianness anyway, so I don't see any perfomance gain by doing this this way. I would suggest accessing _info_ bytewise

Reply:
REG_DATA is a 16-bit register. I think I have to read it with inw(), 2 bytes at a time. I tried to use inb() but I receive incorrect data when doing so (also the driver hangs on polling right after - DRQ bit doesn't get cleared and RDY bit doesn't get set).
I'm not sure how I can make this code portable to big endian systems. Perhaps I should use functions like ntohl() for conversion?

Reply:
No, no, that's not what I meant. I am aware that you need to read 16 bit chunks. I suggest below solution:
info[i] = ((data >> 8) & 0xFF);
info[i + i] = ((data & 0xFF) << 8);

or something like that. So info is only accessed byte-wise.

You'll need to adapt the for loop to this.

Reply:
Ok, I get it now.

Reply:
Fixed.